Hudi Test Suite

n3nash opened the pull request:
- Flexible schema payload generation
- Different types of workload generation such as inserts, upserts etc
- Post process actions to perform validations
- Interoperability of test suite to use HoodieWriteClient and HoodieDeltaStreamer so both code paths can be tested
- Custom workload sequence generator
- Ability to perform parallel operations, such as upsert and compaction
- Simple default implementation for validation of datasets, this is left open to the user of test suite and is free to use any sort of validation that they can provide.

n3nash commented:
Created a new module `hoodie-bench`

bvaradar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/sink/writer/FileSinkWriter.java:
```diff
@@ -0,0 +1,37 @@
+  void open(String path) throws IOException;
+
+  /**
+   * Start a new file sink writer and close the current one
```
Looks like the implementation does not auto-close. Either the javadoc needs to be fixed or the implementation needs to change.

n3nash replied:
fixed

bvaradar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/helpers/DFSTestSuitePathSelector.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * A custom dfs path selector used only for the hudi test suite. To be used only if workload is not run inline.
+ */
+public class DFSTestSuitePathSelector extends DFSPathSelector {
```
From usability perspective, would greatly help if we a bunch of ready to run test showing the power of hoodie-bench in testing. Maybe you already planned for this.  Let me know your thoughts ?

n3nash replied:
Yes, in the following PR I plan to write up atleast 1 test to cover all possible cases supported by hoodie-bench

vinothchandar replied:
I feel we should have one end-end test before we land this.

n3nash commented:
@vinothchandar gentle reminder

n3nash commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/validate/ValidateCount.java:
```diff
@@ -0,0 +1,26 @@
+  private static Logger log = Logger.getLogger(ValidateCount.class);
+
+  @Override
+  public Long validate(HudiTestSuiteConfig cfg,
```
Yes, that will be part of the next PR as a sample validator for the test job.

n3nash commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/workload/WorkloadManager.java:
```diff
@@ -0,0 +1,289 @@
+          case SCHEDULE_COMPACTION:
+            // Can only be done with an instantiation of a new WriteClient hence cannot be done during DeltaStreamer
+            // testing for now
+            // Find the last commit and extra the extra metadata to be passed to the schedule compaction. This is
```
fixed

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/job/action/ValidateAction.java:
```diff
@@ -0,0 +1,18 @@
+
+  private String validationClassName;
+
+  public ValidateAction(String validationClassName) {
```
The operation and action classes can just be a ENUM IMO.. unless you want to add more methods here, it seems superflous to me.

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/job/action/result/JavaRDDResult.java:
```diff
@@ -0,0 +1,10 @@
+
+import org.apache.spark.api.java.JavaRDD;
+
+public class JavaRDDResult extends ActionResult<JavaRDD> {
```
same with these methods? I am unable to clear understand how these classes are going to evolve? if these will be passed to a validate method, then it should be enough to simply templatize that?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/job/action/Action.java:
```diff
@@ -0,0 +1,9 @@
+ * All types of actions that can be tested using the test suite in parallel with workload operations
+ * {@link com.uber.hoodie.integrationsuite.configuration.WorkloadOperationConfig}
+ */
+public enum Action {
```
rename to ActionType?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/job/ThreadPoolService.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Thread pool service for common helper methods
+ */
+public class ThreadPoolService {
```
can we avoid this class by using either https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#invokeAll-java.util.Collection- or parallelStreams?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/sink/writer/SinkWriter.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @param <I> Data type to be written to the sink
+ */
+public interface SinkWriter<I> {
```
I am not sure if we should invest into a hierarchy of classes for this. Can't we use Spark datasources out of box to do this.. There are ones that write avro files into dfs already, same with writing to kafka..

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/job/operation/Operation.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Types of operations supported by Hudi and HudiDeltaStreamer
+ */
+public enum Operation {
```
rename to OperationType?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/converter/InsertToUpdateConverter.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ */
+
+package com.uber.hoodie.integrationsuite.converter;
```
do we need a package for a single class?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/converter/InsertToUpdateConverter.java:
```diff
@@ -0,0 +1,54 @@
+  private final List<String> recordKeyFields;
+  private final int minPayloadSize;
+
+  public InsertToUpdateConverter(String schemaStr, int minPayloadSize, List<String> partitionPathFields,
```
I don't quite follow how we decide this key has been already inserted in a previous batch? do we track all keys inside the dataset somewhere? what am I missing?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/generator/GenericRecordFullPayloadGenerator.java:
```diff
@@ -0,0 +1,232 @@
+  // Used to validate a generic record
+  private final transient GenericData genericData = new GenericData();
+  // Number of more bytes to add based on the estimated full record payload size and min payload size
+  private int numberOfBytesToAdd;
```
so the idea is even if the schema is complex and large we fill up to the specified amount of bytes? if so, do we need this kind of complexity? Can't we just define a bunch of thin and fat records outside centrally and always generate values for each field in the schema? (can support %nulls to test nulls as a config if thats a reason)

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/generator/GenericRecordFullPayloadGenerator.java:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  public GenericRecord getNewPayload() {
+    return convert(baseSchema);
```
how fast is this generation? I tried to use an external avro data generator before and it was just too painfully slow that I had to write a simplified one.

just confirming that this generation is done, persisted to dfs/kafka and then delta streamer reads from sink again? 
I was concerned that if we inline this with the hudi dag, then if this conversion is slow, then it will seem like ingestion/writing is slow.

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/generator/WorkloadGenerator.java:
```diff
@@ -0,0 +1,196 @@
+/**
+ * The workload generator generates all types of workloads (insert, update) for the given configs
+ */
+public class WorkloadGenerator implements Serializable {
```
Rename to DeltaGenerator/DeltaProducer? Seems we are overloading the term Workload? Also plays nicely with DeltaStreamer in the sense, we generate deltas and then the deltastreamer consumes them

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/helpers/AvroReader.java:
```diff
@@ -0,0 +1,56 @@
+  public static JavaRDD<GenericRecord> read(SparkSession sparkSession, String schemaStr, List<String> listOfPaths,
+      Optional<String> structName, Optional<String> nameSpace) {
+
+    Dataset<Row> dataSet = sparkSession.read()
```
follow on to the previous comment.. we can just use sparkSession.write() to write out to the sink?

vinothchandar commented on hoodie-bench/src/main/java/com/uber/hoodie/integrationsuite/helpers/DFSTestSuitePathSelector.java:
```diff
@@ -0,0 +1,74 @@
+    Integer nextBatchId;
+    try {
+      if (lastCheckpointStr.isPresent()) {
+        lastBatchId = Integer.parseInt(lastCheckpointStr.get());
```
main customization seems to be the reading of input, batch by batch? Won't tailing the dfs by modification time as support by deltastreamer not already achieve this?

vinothchandar commented:
@n3nash I thought you were nt working on this. @yanghua  is also working on something similar in HUDI-289 .. Can you share some context?

n3nash commented:
@vinothchandar I've updated the PR. I worked on it a while ago but never upstreamed the local changes. Ran into some issues making Hive queries work which is also fixed. I'll rebase this by tomorrow and hopefully address most of the comments by Monday.
@yanghua Please take a look at this PR when you get a chance and we can follow it up with thoughts/ideas you may have.

yanghua commented:
Thanks for updating this PR. I have tried to pick this PR to my local then fix the conflicts and continue to work based on it. However, I found there are many conflicts so I gave up this thought.

The things what I did in my local is:

- created a new module named `hudi-end-to-end-test`;
- copied your most file of this PR into my local module;
- tried to fix the errors (there are still some errors to be fixed)
- refactored dag package and provided an `ExecutionContext` POJO and unified `execute` method as an abstract method and refactored `DagScheduler`
- understanding parsing yaml file business logic

It's OK about working together.

vinothchandar commented:
Just want to ensure we have a plan going forward.. @yanghua how far along are you? @n3nash can you please work with vino to see how/if both the works can be merged?

bvaradar commented:
@yanghua : I thought your test leverages DistributedTestDataSource and runs for long time and doing validations like counts check. Let me know if this is not the case ?

@n3nash test suite uses controlled orchestration and hive integration (white box like model) to create different scenarios and validate them.

IIUC, it sounds to me these are complementary and we will have uses for both.

yanghua commented:
> @yanghua : I thought your test leverages DistributedTestDataSource and runs for long time and doing validations like counts check. Let me know if this is not the case ?
> 

We are planning to implement this. However, it have not been finished yet.

vinothchandar commented:
@n3nash has a PR up I think #991 . if you have not started on a bunch of work already, we can use that and see how to improve upon?

n3nash commented:
Please follow this PR : https://github.com/apache/incubator-hudi/pull/991. Ran into a lot of errors rebasing this PR since this was opened before the Hoodie -> Hudi rename, hence opened a new PR.

yanghua commented:
> Please follow this PR : #991. Ran into a lot of errors rebasing this PR since this was opened before the Hoodie -> Hudi rename, hence opened a new PR.

OK, Let's move to the new PR #991 , will checkout it and continue to work based on it.

n3nash commented on hoodie-bench/prepareIntegrationSuite.sh:
```diff
@@ -0,0 +1,123 @@
+#!/bin/bash
```
renamed

n3nash commented:
Closing this one in favor of the new one
